kotlin/flatbuffers-kotlin/src/commonMain/kotlin/com/google/flatbuffers/kotlin/JSON.kt

```diff
@@ -0,0 +1,788 @@
+          cursor++
+          if (cursor == limit) break
+          tok = ary[cursor]
+        } while (tok in LETTER_0..LETTER_9)
+      }
+    }
+
+    var exponent = 0
+    // If we find '.' we need to convert to double
+    if (tok == LETTER_DOT) {
+      useDouble = true
+      checkEOF(data, cursor)
+      tok = ary[++cursor]
+      if (tok < LETTER_0 || tok > LETTER_9) {
+        makeError(data, "Invalid Number", tok)
+      }
+      do {
+        // double = double * 10 + (tok - LETTER_0)
+        long = 10 * long + (tok - LETTER_0)
+        digits++
+        --exponent
+        cursor++
+        if (cursor == limit) break
+        tok = ary[cursor]
+      } while (tok in LETTER_0..LETTER_9)
+    }
+
+    // If we find 'e' we need to convert to double
+    if (tok == LETTER_e || tok == LETTER_E) {
+      useDouble = true
+      ++cursor
+      checkEOF(data, cursor)
+      tok = ary[cursor]
+      var negativeExponent = false
+      if (tok == LETTER_MINUS) {
+        ++cursor
+        checkEOF(data, cursor)
+        negativeExponent = true
+        tok = ary[cursor]
+      } else if (tok == LETTER_PLUS) {
+        ++cursor
+        checkEOF(data, cursor)
+        tok = ary[cursor]
+      }
+      if (tok < LETTER_0 || tok > LETTER_9) {
+        makeError(data, "Missing exponent", tok)
+      }
+      var exp = 0
+      do {
+        val digit = tok - LETTER_0
+        exp = 10 * exp + digit
+        ++cursor
+        if (cursor == limit) break
+        tok = ary[cursor]
+      } while (tok in LETTER_0..LETTER_9)
+
+      exponent += if (negativeExponent) -exp else exp
+    }
+
+    if (digits > 17 || exponent < -19 || exponent > 19) {
+      // if the float number is not simple enough
+      // we use language's Double parsing, which is slower but
+      // produce more expected results for extreme numbers.
+      val firstPos = readPos - 1
+      val str = data.getString(firstPos, cursor - firstPos)
+      if (useDouble) {
+        double = str.toDouble()
+        output[key] = double
+      } else {
+        long = str.toLong()
+        output[key] = long
+      }
+    } else {
+      // this happens on single numbers outside any object
+      // or array
+      if (useDouble || exponent != 0) {
+        double = if (long == 0L) 0.0 else long.toDouble() * exponent.pow10()
+        double *= sign
+        output[key] = double
+      } else {
+        long *= sign
+        output[key] = long
+      }
+    }
+    readPos = cursor
+    return if (useDouble) T_FLOAT else T_INT
+  }
+
+  private fun parseString(data: ReadBuffer, key: String?): FlexBufferType {
+    output[key] = readString(data)
+    return T_STRING
+  }
+
+  private fun parseTrue(data: ReadBuffer, key: String? = null): FlexBufferType {
+    checkEOF(data, readPos + 2)
+    // 0x65757274 is equivalent to ['t', 'r', 'u', 'e' ] as a 4 byte Int
+    if (data.getInt(readPos - 1) != 0x65757274) {
+      makeError(data, "Expecting keyword \"true\"", data[readPos])
```

Comment:
see, these are really weird errors.. wouldn't happen if you had general lexical analysis that can parse identifiers etc.

Reply:
Why do you think it is weird? I get the lexical analysis, but it will just move the error around, no?

Reply:
in that if you write `truely` you get `expecting keyword true` instead of `unknown identifier truely` etc. No parser operates that way.